tools/codegen/api/types.py

```diff
@@ -40,15 +51,179 @@ def __str__(self) -> str:
+# represent a TensorOptions.  If you don't care about TensorOptions
+# processing, think of this as just a list of four CppArguments; however
+# if you need to bundle these arguments back into a single
+# TensorOptions, it will be easiest to operate on this struct as a
+# whole.
+@dataclass(frozen=True)
+class CppTensorOptionsArguments:
+    argument: TensorOptionsArguments
+    dtype: CppArgument
+    layout: CppArgument
+    device: CppArgument
+    pin_memory: CppArgument
+
+    # Remove the defaults from each of the constituent arguments
+    # representing the TensorOptions
+    def no_default(self) -> 'CppTensorOptionsArguments':
+        return CppTensorOptionsArguments(
+            argument=self.argument,
+            dtype=self.dtype.no_default(),
+            layout=self.layout.no_default(),
+            device=self.device.no_default(),
+            pin_memory=self.pin_memory.no_default(),
+        )
+
+    # Flatten the TensorOptions into individual CppArguments
+    def explicit_arguments(self) -> Sequence[CppArgument]:
+        return [self.dtype, self.layout, self.device, self.pin_memory]
+
 @dataclass(frozen=True)
 class CppExpr:
     type: str
     expr: str
 
+# A CppSignature represents a single overload in the C++ API.  For
+# any given function schema, there may be multiple CppSignatures
+# corresponding to it, based on how we desugar to C++.  See also
+# CppSignatureGroup.
+@dataclass(frozen=True)
+class CppSignature:
+    # The schema this signature is derived from
+    func: FunctionSchema
+
+    # Enough information about the C++ types to generate a full
+    # C++ type signature for this signature.  I'm not too sure
+    # if these are the right representations, so for now this
+    # is intended to be more abstract.  Prefer using
+    # cpp_grouped_arguments() to access this
+    _cpp_grouped_arguments: Tuple[Union[CppArgument, CppTensorOptionsArguments, CppThisArgument], ...]
+    _cpp_returns_type: str
+
+    # WARNING: This is probably NOT what you want
+    #
+    # Return the flattened argument structure of this signature,
+    # discarding information about grouping.  If you are planning
+    # to translate these arguments into expressions on another
+    # API, you probably want cpp_grouped_arguments instead (which
+    # will handle TensorOptions translations correctly)
+    def cpp_ungrouped_arguments(self) -> Sequence[CppArgument]:
+        return [sub_a for a in self._cpp_grouped_arguments for sub_a in a.explicit_arguments()]
+
+    # Return the grouped argument structure of this signature.  This
+    # preserves high-level structure of the arguments so you may
+    # find it easier to do translations working with this
+    # representation.
+    def cpp_grouped_arguments(self) -> Sequence[Union[CppArgument, CppTensorOptionsArguments, CppThisArgument]]:
+        return self._cpp_grouped_arguments
+
+    # Render the C++ declaration for this signature
+    def decl(self) -> str:
```

Comment:
hm this doesn't specify if it returns the grouped or the ungrouped declaration. I think this points to general issues with the data structure representation. We should probably make that difference and the transformation that is happening here more clear. Maybe have two CppSignature classes, one for grouped and one for ungrouped?

After reading through the code below, it seems that you have one CppSignature class but two instances, one for grouped and one for ungrouped. And in the ungrouped instance, the `_cpp_grouped_arguments` member holds the ungrouped arguments. This is very confusing.

Reply:
This is the correct interpretation. I'm happy to rename the `_cpp_grouped_arguments` member, the name here is mostly based off of what the types say is possible. I agree it is possible to have two CppSignature classes, one for grouped and another for ungrouped, but this will increase the boilerplate. I can give it a try though if you want.

Reply:
ok after some offline discussion, it turned out I misunderstood the terms in your PR. I thought "grouped signature" == "non-faithful signature" because the non-faithful signature groups tensor options arguments into one TensorOptions object...but actually, non-faithful signatures don't even use the grouping logic and just have one CppArgument for each argument (one of them being a TensorOptions). Only for faithful signatures, the difference between a "grouped signature" and "ungrouped signature" comes into play and both always represent the faithful signature, just at different abstraction layers.